### PR TITLE
:lock: prevent users to buy the same ticket twice

### PR DESCRIPTION
**Pull request by antoineeudes**

### CODE REVIEW

File: functions/payment.js [hunk 1]
```diff
@@ -98,10 +98,25 @@ exports.createPaymentIntent = functions.https.onCall(async (data, context) => {
       "venue, event or return url data missing"
     );
   }
+
+  const ticketsBoughtByUserForThisEvent = await admin
+    .firestore()
+    .collection("purchases")
+    .where("userId", "==", context.auth.token.user_id)
+    .where("eventId", "==", data.eventId)
+    .where("status", "in", ["COMPLETE"])
```
Comment (maxb94) [inline functions/payment.js:107]: what happens if ticket status is "PENDING" | "CONFIRMATION_FROM_STRIPE_PENDING"? Will the user be able to rebuy the ticket accidentaly?

File: functions/payment.js [hunk 2]
```diff
@@ -156,8 +217,17 @@ exports.webhooks = functions.https.onRequest(async (request, res) => {
...
-  }
+  } else if (event.type === "charge.failed") {
+    const charge = event.data.object;
 
+    await admin
+      .firestore()
+      .collection(PURCHASE_TABLE)
+      .doc(charge.payment_intent)
+      .update({
+        status: "FAILED",
```
Comment (maxb94) [inline functions/payment.js:228]: add FAILED to Purchase type

File: functions/payment.js [hunk 3]
```diff
@@ -125,6 +140,52 @@ exports.createPaymentIntent = functions.https.onCall(async (data, context) => {
...
+      .get()
+  ).data();
+
+  if (!["COMPLETE", "FAILED"].includes(documentToUpdate.status)) {
+    await admin
+      .firestore()
+      .collection(PURCHASE_TABLE)
+      .doc(data.paymentIntentId)
+      .update({
+        status: "CONFIRMATION_FROM_STRIPE_PENDING",
```
Comment (maxb94) [inline functions/payment.js:182]: when do we set status to PENDING. What's the difference between PENDING and CONFIRMATION_FROM...?

File: src/types/Purchase.ts [hunk 4]
```diff
@@ -1,6 +1,13 @@
 export interface Purchase {
   eventId: string;
   userId: string;
-  status: "PENDING" | "COMPLETE" | "CONFIRMATION_FROM_STRIPE_PENDING";
+  // when the user opens the PaymentForm, the status is PENDING
```
Comment (maxb94) [inline src/types/Purchase.ts:4]: nice

File: functions/payment.js [hunk 5]
```diff
@@ -125,6 +140,52 @@ exports.createPaymentIntent = functions.https.onCall(async (data, context) => {
...
+      "invalid-argument",
+      "venue, event or return url data missing"
+    );
+  }
+
+  const documentToUpdate = (
+    await admin
+      .firestore()
+      .collection(PURCHASE_TABLE)
+      .doc(data.paymentIntentId)
```
Comment (mathieud) [inline functions/payment.js:172]: You could use the stripe session id (which is the document key) here I think (the frontend has access to it, it's returned by the createIntent).

File: src/components/organisms/PaymentModal/PaymentForm.tsx [hunk 6]
```diff
@@ -105,6 +105,12 @@ const PaymentForm: React.FunctionComponent<PropsType> = ({
       setErrorMessage(result.error.message);
     } else if (result.paymentIntent?.status === "succeeded") {
       setIsPaymentSuccess(true);
+      await firebase.functions().httpsCallable("payment-confirmPaymentIntent")({
+        venueId: venueId,
```
Comment (mathieud) [inline src/components/organisms/PaymentModal/PaymentForm.tsx:109]: Don't think we need anything except the payment intent here

**Pull request merged**